File: 2021/clojure/src/day02.clj
(ns day02
  (:require [clojure.string :as str]))

(defrecord Instruction [direction value])
(defrecord Position [horizontal depth aim])

(defn parse-instruction [line]
  (let [[direction value] (str/split line #" ")]
    (Instruction. (keyword direction) (Integer/parseInt value))))

(def input
  (->> "data/day02.txt"
       slurp
       str/split-lines
       (map parse-instruction)))

(defn next-position [position instruction]
  (let [horizontal (:horizontal position)
        depth (:depth position)
        direction (:direction instruction)
        value (:value instruction)]
    (case direction
      :forward (assoc position :horizontal (+ horizontal value))
      :down (assoc position :depth (+ depth value))
      :up (assoc position :depth (- depth value)))))

(defn next-position-2 [position instruction]
  (let [horizontal (:horizontal position)
        depth (:depth position)
        aim (:aim position)
        direction (:direction instruction)
        value (:value instruction)]
    (case direction
      :forward (assoc position
                      :horizontal (+ horizontal value)
                      :depth (+ depth (* aim value)))
      :down (assoc position :aim (+ aim value))
      :up (assoc position :aim (- aim value)))))

(defn answer [position]
  (* (:horizontal position) (:depth position)))

(defn part1 []
  (->> input
       (reduce next-position (Position. 0 0 0))
       answer))

(defn part2 []
  (->> input
       (reduce next-position-2 (Position. 0 0 0))
       answer))

(do (println "Part 1:" (part1))
    (println "Part 2:" (part2)))

File: 2021/python/aoc2021/day02.py
from dataclasses import dataclass

from .utils import read_lines_from_file


@dataclass
class Instruction:
    value: int


class Forward(Instruction):
    ...


class Up(Instruction):
    ...


class Down(Instruction):
    ...


def parse_instructions(data: list[str]) -> list[Instruction]:
    instructions: list[Instruction] = []
    for row in data:
        match row.split():
            case ["forward", value]:
                instructions.append(Forward(int(value)))
            case ["up", value]:
                instructions.append(Up(int(value)))
            case ["down", value]:
                instructions.append(Down(int(value)))
    return instructions


def part_1(instructions: list[Instruction]) -> int:
    h = depth = 0
    for instruction in instructions:
        match instruction:
            case Forward(value):
                h += value
            case Up(value):
                depth -= value
            case Down(value):
                depth += value
    return h * depth


def part_2(instructions: list[Instruction]) -> int:
    h = depth = aim = 0
    for instruction in instructions:
        match instruction:
            case Forward(value):
                h += value
                depth += aim * value
            case Up(value):
                aim -= value
            case Down(value):
                aim += value
    return h * depth


def main():
    data = read_lines_from_file("day02.txt")
    instructions = parse_instructions(data)
    print("Part 1:", part_1(instructions))
    print("Part 2:", part_2(instructions))


if __name__ == "__main__":
    main()
